File: plugins/delay/tape-delay/tape-delay.js
// Tape Delay — single-tap feedback delay with a one-pole low-pass on
// each repeat (the "tone" knob). Each subsequent echo loses high-end
// just like analogue tape, so a long feedback time slowly degrades
// into a warm low-frequency soup instead of harsh repeating spikes.

function TapeDelayProcessor() {
  // Pre-allocate buffer for the max delay (1.5 s @ assumed-high sample rate).
  // 96 kHz max gives 144 000 samples; round up.
  this.buf = new Array(150000);
  for (var k = 0; k < this.buf.length; k++) this.buf[k] = 0;
  this.bufLen = this.buf.length;
  this.writeIdx = 0;
  this.lpState = 0;
}
TapeDelayProcessor.parameterDescriptors = [
  { name: 'time',     label: 'Time (ms)', defaultValue: 350, minValue: 5,  maxValue: 1500 },
  { name: 'feedback', label: 'Feedback',  defaultValue: 0.4, minValue: 0,  maxValue: 0.95 },
  { name: 'tone',     label: 'Tone',      defaultValue: 0.5, minValue: 0,  maxValue: 1 },
  { name: 'mix',      label: 'Mix',       defaultValue: 0.4, minValue: 0,  maxValue: 1 }
];
TapeDelayProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0];
  var output = outputs[0][0];
  if (!input || !output) return true;
  var timeMs = parameters.time[0];
  var feedback = parameters.feedback[0];
  var tone = parameters.tone[0];
  var mix = parameters.mix[0];
  var delaySamples = Math.max(1, Math.floor(sampleRate * timeMs / 1000));
  // One-pole low-pass coefficient: tone=1 → fully open, tone=0 → muffled.
  var lpA = 0.05 + 0.93 * tone;

  for (var i = 0; i < input.length; i++) {
    var readIdx = this.writeIdx - delaySamples;
    while (readIdx < 0) readIdx += this.bufLen;
    var delayed = this.buf[readIdx];

    // Filter the wet path for the next round (in-feedback tone shaping).
    this.lpState = this.lpState + lpA * (delayed - this.lpState);
    var fbSample = this.lpState;

    // Write input + feedback back into the line.
    this.buf[this.writeIdx] = input[i] + fbSample * feedback;
    this.writeIdx = (this.writeIdx + 1) % this.bufLen;

    output[i] = input[i] * (1 - mix) + fbSample * mix;
  }
  return true;
};

registerProcessor('tape-delay', TapeDelayProcessor);

File: plugins/delay/tape-delay/TapeDelay.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Tape Delay (native port) — single-tap feedback delay with a one-pole
// low-pass on each repeat ("tone"). Each echo loses high-end like real
// analogue tape, so long feedback degrades into warm low-freq soup
// rather than harsh repeating spikes.
public final class TapeDelay implements VocalMonitorNativePlugin {
    private float[] buf;
    private int bufLen;
    private int writeIdx = 0;
    private float lpState = 0f;
    private int sampleRate = 44100;
    private float time = 350f, feedback = 0.4f, tone = 0.5f, mix = 0.4f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        bufLen = 150000;
        buf = new float[bufLen];
        writeIdx = 0;
        lpState = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "time", "feedback", "tone", "mix" }; }
    @Override public float parameterMin(String n) {
        if ("time".equals(n)) return 5f;
        return 0f;
    }
    @Override public float parameterMax(String n) {
        switch (n) {
            case "time":     return 1500f;
            case "feedback": return 0.95f;
            default:         return 1f;
        }
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "time":     return 350f;
            case "feedback": return 0.4f;
            case "tone":     return 0.5f;
            case "mix":      return 0.4f;
            default:         return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "time":     return "Time (ms)";
            case "feedback": return "Feedback";
            case "tone":     return "Tone";
            case "mix":      return "Mix";
            default:         return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "time":     time = v; break;
            case "feedback": feedback = v; break;
            case "tone":     tone = v; break;
            case "mix":      mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final int delaySamples = Math.max(1, (int) Math.floor(sampleRate * time / 1000f));
        final float lpA = 0.05f + 0.93f * tone;
        final float fb = feedback;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final float[] b = buf;
        final int bL = bufLen;
        int w = writeIdx;
        float lp = lpState;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            int r = w - delaySamples;
            while (r < 0) r += bL;
            float delayed = b[r];
            lp = lp + lpA * (delayed - lp);
            b[w] = input[i] + lp * fb;
            w++; if (w >= bL) w = 0;
            output[i] = input[i] * dry + lp * mixLocal;
        }
        writeIdx = w;
        lpState = lp;
    }
}
